rust-lang/cargo
# Basic implementation for `cargo install --dry-run`

### What does this PR try to resolve?

This is an attempt at providing a starting implementation for `cargo install --dry-run` discussed in #11123. The precise objective is more to have something that works now rather than a much more complete version requiring time and complexity-facing skills I don't currently have.

Indeed, https://github.com/rust-lang/cargo/issues/11123#issuecomment-1910469992 suggested to descend into `InstallablePackage::install_one` and its inner guts. That would enable getting much more accurate information relatively to the usual non-dry-run operation. However, it looked a bit too complicated for a reviewable first step and also not too orthogonal with the `build-plan` and `unit-graph` features. The plan I adopted in this attempt is therefore to:
 * only handle the CLI flag in `cargo_install::install`;
 * basically only use the information provided by `InstallablePackage::new` and other easily-retrievable ones;
 * put itself between that and the start of the compilation to cut it entirely short;
 * display status messages indicating which very broad actions would be performed, but not much more.

As a quick preview:
```console
$ cargo run --bin cargo -- -Z unstable-options install --dry-run cargo-liner gitoxide bat git-delta
    Finished dev [unoptimized + debuginfo] target(s) in 0.14s
     Running `target/debug/cargo -Z unstable-options install --dry-run cargo-liner gitoxide bat git-delta`
    Updating crates.io index
     Ignored package `cargo-liner v0.6.0` is already installed, use --force to override
    Updating crates.io index
  Downloaded gitoxide v0.33.0
  Downloaded 1 crate (75.5 KB) in 0.18s
     Ignored package `bat v0.24.0` is already installed, use --force to override
     Summary dry-run
 Destination /home/pauldance/.cargo/bin
     Install gitoxide: to 0.33.0 using registry `crates-io`
     Upgrade git-delta: from 0.16.5 to 0.17.0 using registry `crates-io`
  Up-to-date cargo-liner
  Up-to-date bat
```

### How should we test and review this PR?

Several tests have been added in order to check that the usual combinations of the flag with `--git`, `--path` and installation states behave as expected and print the desired outputs. See the end of `tests/testsuite/install.rs` for more details about this.

Review should be made easier using the few provided commits:
 * CLI change.
 * Implementation.
 * Tests.
 * Docs.

In particular, I might not be too fond of the wording I chose here and there. The way the information is displayed in a split manner using several new status kinds might not be the best either considering the rest of Cargo's output. Feel free to correct me on any of this.

### Additional information

This is very much a first implementation that could evolve in the near future depending on the precision of the emulation that is desirable. As-is, most of the operation is entirely avoided and only informational messages are printed instead. This feels to me enough for a first step. Being more precise would require emulating the compilation jobs or even their inner operations, at the cost of additional complexity.

Obviously, any feedback is welcome, especially if it is about the next steps potentially requiring the mentioned accuracy or if these should be done now instead.
